android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -13,291 +20,586 @@
+  public static void throwIfInsufficientLevel(@NonNull final CipherStorage storage,
+                                              @NonNull final SecurityLevel level)
+    throws CryptoFailedException {
+    if (storage.securityLevel().satisfiesSafetyThreshold(level)) {
+      return;
+    }
 
-            promise.resolve(credentials);
-        } catch (KeyStoreAccessException e) {
-            Log.e(KEYCHAIN_MODULE, e.getMessage());
-            promise.reject(E_KEYSTORE_ACCESS_ERROR, e);
-        } catch (CryptoFailedException e) {
-            Log.e(KEYCHAIN_MODULE, e.getMessage());
-            promise.reject(E_CRYPTO_FAILED, e);
-        }
+    throw new CryptoFailedException(
+      String.format(
+        "Cipher Storage is too weak. Required security level is: %s, but only %s is provided",
+        level.name(),
+        storage.securityLevel().name()));
+  }
+
+  @Nullable
+  private CipherStorage getCipherStorageByName(@NonNull final String cipherStorageName) {
+    return cipherStorageMap.get(cipherStorageName);
+  }
+
+  /** True - if fingerprint hardware available and configured, otherwise false. */
+  /* package */ boolean isFingerprintAuthAvailable() {
+    return DeviceAvailability.isFingerprintAuthAvailable(getReactApplicationContext());
+  }
+
+  /** Is secured hardware a part of current storage or not. */
+  /* package */ boolean isSecureHardwareAvailable() {
+    try {
+      return getCipherStorageForCurrentAPILevel().supportsSecureHardware();
+    } catch (CryptoFailedException e) {
+      return false;
     }
+  }
 
-    private DecryptionResult decryptCredentials(String service, CipherStorage currentCipherStorage, ResultSet resultSet) throws CryptoFailedException, KeyStoreAccessException {
-        if (resultSet.cipherStorageName.equals(currentCipherStorage.getCipherStorageName())) {
-            // The encrypted data is encrypted using the current CipherStorage, so we just decrypt and return
-            return currentCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
-        }
+  /** Resolve storage to security level it provides. */
+  @NonNull
+  private SecurityLevel getSecurityLevel() {
+    try {
+      final CipherStorage storage = getCipherStorageForCurrentAPILevel();
 
-        // The encrypted data is encrypted using an older CipherStorage, so we need to decrypt the data first, then encrypt it using the current CipherStorage, then store it again and return
-        CipherStorage oldCipherStorage = getCipherStorageByName(resultSet.cipherStorageName);
-        // decrypt using the older cipher storage
+      if (!storage.securityLevel().satisfiesSafetyThreshold(SECURE_SOFTWARE)) {
+        return ANY;
+      }
 
-        DecryptionResult decryptionResult = oldCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
-        // encrypt using the current cipher storage
+      if (isSecureHardwareAvailable()) {
+        return SECURE_HARDWARE;
```

Comment:
To follow code style used in this library:
```suggestion
        return SecurityLevel.SECURE_HARDWARE;
```